### PR TITLE
Allow basic support for MySQL year type

### PR DESCRIPTION
**Pull request by dereuromark**
Follow up on https://github.com/cakephp/cakephp/pull/9654

I now confirmed with a real test application that it works together with a custom [YearType](https://github.com/dereuromark/cakephp-shim/blob/master/src/Database/Type/YearType.php).

This patch is required to properly store the values posted on the one side, and to output format it as Year dropdown on the other side for form display.

### CODE REVIEW

**Comment by thinkingmedia** [comment]
I've been collecting together things related to the database as a project for 3.4.

https://github.com/cakephp/cakephp/projects/2

Want to include this for 3.4?

File: tests/TestCase/View/Widget/DateTimeWidgetTest.php [hunk 1]
```diff
@@ -1198,6 +1198,28 @@ public function testRenderMeridianWidget()
...
+            'hour' => false,
+            'minute' => false,
+            'second' => false,
+        ], $this->context);
+
+        $expected = '<select name="date[year]"><option value="';
+        $this->assertContains($expected, $result);
+
+        $expected = '<option value="2009" selected="selected">2009</option></select>';
+        $this->assertContains($expected, $result);
```
Comment (markstory) [inline tests/TestCase/View/Widget/DateTimeWidgetTest.php]: Should you ensure other inputs are not present?

Comment (dereuromark) [inline tests/TestCase/View/Widget/DateTimeWidgetTest.php]: I did with the first and second assert together afaik.

**Comment by dereuromark** [comment]
@thinkingmedia Feel free to pull this into 3.4, if you want to. I am fine either way.

**Comment by thinkingmedia** [comment]
@dereuromark thanks, done.

File: src/Database/Schema/MysqlSchema.php [hunk 2]
```diff
@@ -92,7 +92,7 @@ protected function _convertColumn($column)
             $precision = (int)$precision;
         }
 
-        if (in_array($col, ['date', 'time', 'datetime', 'timestamp'])) {
+        if (in_array($col, ['date', 'time', 'datetime', 'timestamp', 'year'])) {
             return ['type' => $col, 'length' => null];
```
Comment (markstory) [inline src/Database/Schema/MysqlSchema.php]: We don't have a concrete type for this, how will the ORM handle mapping this type?

Comment (dereuromark) [inline src/Database/Schema/MysqlSchema.php]: You can Hook in your own Type to handle it. See my previous pr and comments.

Comment (thinkingmedia) [inline src/Database/Schema/MysqlSchema.php]: what's wrong with just mapping it to integer?

Comment (markstory) [inline src/Database/Schema/MysqlSchema.php]: But the 'abstract' type will be 'year', will that convert back into the correct `CREATE TABLE` statement?

Comment (markstory) [inline src/Database/Schema/MysqlSchema.php]: If a table has a `YEAR` column and we reflect it into a `TableSchema` object, can we then generate the correct SQL again?

Comment (dereuromark) [inline src/Database/Schema/MysqlSchema.php]: I might need some help to answer that correctly.

Comment (markstory) [inline src/Database/Schema/MysqlSchema.php]: Once you have a schema object from reflection, you can use `$schema->createSql($table->connection())` to get the generated SQL.

**Pull request closed without merging**